File: Arrays1/lesson1/Javascript/04_manual_array.js
class CustomArray{
    constructor(){
        this.data = {};
        this.length = 0;
    }

    //Access
    get(index){
        return this.data[index]
    }

    //Update
    set(index, value){
        if(index < this.length){
            this.data[index] = value;
        }
    }

    //Insert at end
    push(value){
        this.data[this.length] = value;
        this.length ++;
    }

    //Insert at beginning
    unshift(value){
        for(let i = this.length; i > 0; i --){
            this.data[i] = this.data[i - 1]
        }
        this.data[0] = value;
        this.length ++;
    }

    //Delete at end
    pop(){
        if(this.length === 0) return undefined;
        const last = this.data[this.length - 1];
        delete this.data[this.length -1];
        this.length --;
        return last;
    }

    //Delete at beginning
    shift(){
        if(this.length === 0) return undefined;
        const first = this.data[0];
        for(let i = 0; i < this.length -1 ; i++){
            this.data[i] = this.data[i + 1];
        }
        delete this.data[this.length - 1];
        this.length --;
        return first;
    }

    //Linear search
    indexof(value){
        for(let i = 0; i < this.length; i++){
            if(this.data[i] === value) return i;
        }
        return -1;
    }
}

File: Arrays1/lesson1/Python/04_manual_array.py
class CustomArray:
    def __init__(self):
        self.data = {}
        self.length = 0

    #Access
    def get(self, index):
        return self.data.get(index, None)
    
    #Update
    def set(self, index, value):
        if index < self.length:
            self.data[index] = value
    
    #Insert at end
    def push(self, value):
        self.data[self.length] = value
        self.length += 1

    #Insert at beginning
    def unshift(self, value):
        for i in range(self.length, 0, -1):
            self.data[i] = self.data[i - 1]
        self.data[0] = value
        self.length += 1

    #Delete at end
    def pop(self):
        if self.length == 0:
            return None
        last = self.data[self.length - 1]
        del self.data[self.length -1]
        self.length -= 1
        return last
    
    #Delete at beginning
    def shift(self):
        if self.length == 0:
            return None
        first = self.data[0]
        for i in range(self.length - 1):
            self.data[i] = self.data[i + 1]
        del self.data[self.length - 1]
        self.length -= 1
        return first
    
    #linear search
    def index_of(self, value):
        for i in range(self.length):
            if self.data[i] == value:
                return i
        return  -1

arr = CustomArray()
arr.push(10)
arr.push(20)
arr.unshift(5)
print(arr.get(1)) # 10
print(arr.index_of(20)) # 2
arr.pop()
arr.shift()
print(arr.data)